langchain/src/agents/agent.ts

```diff
@@ -159,14 +160,99 @@ function isAgentAction(input: unknown): input is AgentAction {
   return !Array.isArray(input) && (input as AgentAction)?.tool !== undefined;
 }
 
+export function isRunnableAgent(x: BaseAgent) {
+  return (
+    (x as RunnableMultiActionAgent | RunnableSingleActionAgent).runnable !==
+    undefined
+  );
+}
+
 /**
- * Class representing a single action agent which accepts runnables.
- * Extends the BaseMultiActionAgent class and provides methods for
+ * Class representing a single-action agent powered by runnables.
+ * Extends the BaseSingleActionAgent class and provides methods for
  * planning agent actions with runnables.
  */
-export class RunnableAgent extends BaseMultiActionAgent {
-  protected lc_runnable = true;
+export class RunnableSingleActionAgent extends BaseSingleActionAgent {
+  lc_namespace = ["langchain", "agents", "runnable"];
+
+  runnable: Runnable<
+    ChainValues & { steps: AgentStep[] },
+    AgentAction | AgentFinish
+  >;
+
+  get inputKeys(): string[] {
+    return [];
+  }
+
+  /**
+   * Whether to stream from the runnable or not.
+   * If true, the underlying LLM is invoked in a streaming fashion to make it
+   * possible to get access to the individual LLM tokens when using
+   * `streamLog` with the Agent Executor. If false then LLM is invoked in a
+   * non-streaming fashion and individual LLM tokens will not be available
+   * in `streamLog`.
+   *
+   * Note that the runnable should still only stream a single action or
+   * finish chunk.
+   */
+  streamRunnable = true;
+
+  defaultRunName = "RunnableAgent";
+
+  constructor(fields: RunnableSingleActionAgentInput) {
+    super(fields);
+    this.runnable = fields.runnable;
+    this.defaultRunName = fields.defaultRunName ?? this.defaultRunName;
+    this.streamRunnable = fields.streamRunnable ?? this.streamRunnable;
+  }
+
+  async plan(
+    steps: AgentStep[],
+    inputs: ChainValues,
+    callbackManager?: CallbackManager,
+    config?: RunnableConfig
+  ): Promise<AgentAction | AgentFinish> {
+    const combinedInput = { ...inputs, steps };
+    const combinedConfig = patchConfig(config, {
+      callbacks: callbackManager,
+      runName: this.defaultRunName,
+    });
+    if (this.streamRunnable) {
+      const stream = await this.runnable.stream(combinedInput, combinedConfig);
+      let finalOutput: AgentAction | AgentFinish | undefined;
+      for await (const chunk of stream) {
+        if (finalOutput === undefined) {
+          finalOutput = chunk;
+        } else {
+          throw new Error(
```

Comment:
This is saying that if the runnable streams more than one chunk we'll throw? I'm assuming LLMs will stream back multiple chunks and not one.

Reply:
All the example `RunnableSequence`s powering the agents will gather the output of the LLM and parse it, returning it in one chunk. Streaming is done using intermediate steps.